File: 84/main.java
class Pair {
	public int first;// ht
	public int second;// idx

	public Pair(int a, int b) {
		first = a;
		second = b;
	}
}

class Solution {
	public int largestRectangleArea(int[] heights) {

		Stack<Pair> s = new Stack<Pair>();
		// stack<pair<int, int>> s;

		int n = heights.length;
		// vector<int> left(n), right(n);
		int[] left = new int[n];
		int[] right = new int[n];

		//left - NSL
		for (int i = 0; i < n; i++) {
			if (s.empty())
				left[i] = -1;
			else if (heights[i] > s.peek().first)
				left[i] = s.peek().second;
			else {
				while (!s.empty() && heights[i] <= s.peek().first)
					s.pop();
				if (s.empty())
					left[i] = -1;
				else
					left[i] = s.peek().second;
			}

			s.push(new Pair(heights[i], i));
		}


		while (!s.empty())
			s.pop();

		//right - 1 3 4 2 1 - NSR
		for (int i = n - 1; i >= 0; i--) {
			if (s.empty())
				right[i] = n;
			else if (heights[i] > s.peek().first)
				right[i] = s.peek().second;
			else {
				while (!s.empty() && heights[i] <= s.peek().first)
					s.pop();
				if (s.empty())
					right[i] = n;
				else
					right[i] = s.peek().second;
			}

			s.push(new Pair(heights[i], i));
		}

		int mx = 0;
		for (int i = 0; i < n; i++) {
			mx = Math.max(mx, heights[i] * (right[i] - left[i] - 1));
		}

		return mx;
	}
}

File: 84/main.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        s = [] # [(ht, idx)]

        n = len(heights)
        left = [0] * n
        right = [0] * n

        # left NSL
        for i in range(n):
            if len(s) == 0:
                left[i] = -1
            elif heights[i] > s[-1][0]:
                left[i] = s[-1][1]
            else:
                while len(s) != 0 and heights[i] <= s[-1][0]:
                    s.pop()
                if len(s) == 0:
                    left[i] = -1
                else:
                    left[i] = s[-1][1]

            s.append((heights[i], i))

        s = []
        # right NSR
        for i in range(n-1, -1, -1):
            if len(s) == 0:
                right[i] = n
            elif heights[i] > s[-1][0]:
                right[i] = s[-1][1]
            else:
                while len(s) != 0 and heights[i] <= s[-1][0]:
                    s.pop()
                if len(s) == 0:
                    right[i] = n
                else:
                    right[i] = s[-1][1]

            s.append((heights[i], i))   

        mx = 0

        for i in range(n):
            mx = max(mx, heights[i] * (right[i] - left[i] - 1))

        return mx
